### PR TITLE
[DO NOT REVIEW]feat: body validation

### PR DESCRIPTION
**Pull request by jannyHou**
related to https://github.com/strongloop/loopback-next/issues/118

### CODE REVIEW

**Comment by bajtos** [comment]
> - Dereference the schema object from openapiSpec.components.schemas
> - end to end test

Done in 62ea12a0. I updated the pull request description with additional tasks and also things out of scope.

Next, I'll review the changes and look for places that need cleanup and/or better test coverage.

File: packages/openapi-v3/src/controller-spec.ts [hunk 1]
```diff
@@ -178,16 +178,34 @@ function resolveControllerSpec(constructor: Function): ControllerSpec {
         if (!spec.components.schemas) {
           spec.components.schemas = {};
         }
+        if (p.name in spec.components.schemas) {
+          // Preserve user-provided definitions
```
Comment (bajtos) [inline packages/openapi-v3/src/controller-spec.ts]: We don't have any test for this branch in `openapi-v3` testsuite. Should we add one?

File: packages/openapi-v3/src/controller-spec.ts [hunk 2]
```diff
@@ -178,16 +178,34 @@ function resolveControllerSpec(constructor: Function): ControllerSpec {
...
+          continue;
+        }
         const jsonSchema = getJsonSchema(p);
         const openapiSchema = jsonToSchemaObject(jsonSchema);
+        const outputSchemas = spec.components.schemas;
         if (openapiSchema.definitions) {
           for (const key in openapiSchema.definitions) {
-            spec.components.schemas[key] = openapiSchema.definitions[key];
+            // Preserve user-provided definitions
+            if (key in outputSchemas) continue;
```
Comment (bajtos) [inline packages/openapi-v3/src/controller-spec.ts]: This is implicitly tested by an acceptance test in `@loopback/rest`. Should we add a more focused test to `openapi-v3` too?

**Comment by bajtos** [comment]
@jannyHou today, I cleaned up the implementation a bit and added few more tests, both at validation and unit level. As far as I am concerned, this pull request is good to be landed. Could you please review the final patch, perhaps get other team members to review it too, address any review feedback and land this pull request?

I updated the pull request description with the next steps to work on after this PR is landed:

- Edge cases (array, object)
- More descriptive error object including machine-readable details
- Nice to have: unit-tests in openapi-v3 to verify that user-provided definitions are preserved 

And also updated the list of things that are out of scope of DP3:
- Support for multiple entries in requestBody.content map (e.g. application/json vs. application/xml)
- Support for arbitrary $ref references (see #435)
- Reusing (pre)compiled AJV validators for better performance

Could you please create follow-up issues where it makes sense? For example, I am not sure how to approach multiple content-type entries - do we actually support them in other places of the REST layer? Maybe we need an issue to address this aspect of OpenAPI spec holistically, not only for validation.

**Comment by bajtos** [comment]
@jannyHou last but not least - please preserve my authorship when cleaning up the git commit history, see https://help.github.com/articles/creating-a-commit-with-multiple-authors/

File: packages/rest/src/validation/request-body.validator.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+function getRequestBodySchema(
+  requestBodySpec: RequestBodyObject | undefined,
+  globalSchemas: SchemasObject,
+): SchemaObject | undefined {
+  if (!requestBodySpec) return;
+
+  const content = requestBodySpec.content;
+  // FIXME(bajtos) we need to find the entry matching the content-type
```
Comment (jannyHou) [inline packages/rest/src/validation/request-body.validator.ts]: 👍 Good catch, doesn't look a big code change, opened an issue to track it: https://github.com/strongloop/loopback-next/issues/1494
Will try to address it in the next PR(in a reasonable time)

File: packages/rest/src/validation/request-body.validator.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  return resolveSchemaReference(schema.$ref, globalSchemas);
+}
+
+function resolveSchemaReference(ref: string, schemas: SchemasObject) {
+  // A temporary solution for resolving schema references produced
+  // by @loopback/repository-json-schema. In the future, we should
+  // support arbitrary references anywhere in the OpenAPI spec.
+  // See https://github.com/strongloop/loopback-next/issues/435
```
Comment (jannyHou) [inline packages/rest/src/validation/request-body.validator.ts]: 👍

File: packages/rest/test/helpers.ts [hunk 5]
```diff
@@ -25,3 +30,16 @@ export function createUnexpectedHttpErrorLogger(
     );
   };
 }
+
+export function aBodySpec(
```
Comment (jannyHou) [inline packages/rest/test/helpers.ts]: 👍 I love the idea to create a function generating a body spec, and we can extend the function to allow multiple contents in story https://github.com/strongloop/loopback-next/issues/1494

File: packages/rest/test/unit/request-body.validator.test.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+function verifyValidationRejectsInputWithError(
+  errorMatcher: Error | RegExp | string,
+  body: object | null,
+  spec: RequestBodyObject | undefined,
+  schemas?: SchemasObject,
+) {
+  // workaround for Function.prototype.bind not preserving argument types
+  function validateRequestBodyWithBoundArgs() {
+    validateRequestBody(body, spec, schemas);
```
Comment (jannyHou) [inline packages/rest/test/unit/request-body.validator.test.ts]: @bajtos Any reason why we don't run the unit tests with a mock request but testing `validateRequestBody` directly? Is it because the e-2-e test takes the responsibility of testing a real request?

File: packages/rest/src/parser.ts [hunk 7]
```diff
@@ -65,13 +76,15 @@ async function loadRequestBodyIfNeeded(
...
   const contentType = getContentType(request);
+  debug('Loading request body with content type %j', contentType);
   if (contentType && !/json/.test(contentType)) {
     throw new HttpErrors.UnsupportedMediaType(
       `Content-type ${contentType} is not supported.`,
     );
   }
 
   return await parseJsonBody(request).catch((err: HttpError) => {
+    debug('Cannot parse request body', err);
```
[the remark this answers is not in the document]
Comment (bajtos) [reply packages/rest/src/parser.ts]: I intentionally didn't use `%j` because I want the error to be formatted using `util.inspect`, not `JSON.stringify`.

I think we should use `%s` and `inspect(err)` as we are already doing in other places?

[the remark this answers is not in the document]
Comment (jannyHou) [reply packages/rest/src/parser.ts]: I will change it in https://github.com/strongloop/loopback-next/pull/1495

**Comment by jannyHou** [comment]
@bajtos I was trying to address "More descriptive error object including machine-readable details" in the next PR, but didn't find any clue from [AJV doc](https://github.com/epoberezkin/ajv), could you elaborate more about the "machine-readable details"? Thanks!

**Comment by bajtos** [comment]
Here is the end-user experience I am looking for:

Imagine you are building a single-page web application with a registration form where the user enters email, password, name, etc. When they submit the form, the page makes a REST request to LB4 backend. When the request fails because of a validation error, the page needs to tell user what went wrong and how to fix the filled data. To do so, the page needs to know:
 - Which model properties are not valid, so that corresponding input forms can be highlighted (e.g. `email`, `password`).
 - Validation error to display to the user for each input form. It should be easy to localize these errors, therefore we need an error code (e.g. `code: "required"` for a missing required field) in addition to English message (e.g. "email is required" - whatever is returned back by AJV).

See how LB3.x/Juggler validation works and the example of a ValidationError shown here: https://loopback.io/doc/en/lb3/Validating-model-data.html#localizing-validation-messages

From what I remember, AJV provides additional structured data about the validation errors, we just need to find out where it is stored (I think it's `ajv.errors`) and how to present it to REST clients in the HTTP response. See https://github.com/epoberezkin/ajv#validation-errors for more details about the format of the error object returned by AJV, you may need to inspect the actual object at runtime to build a better understanding.

**Comment by David-Mulder** [comment]
@virkt25 Do note: Most sites already expose whether an a username/email is already registered on the registration page, so 

> bad practice to point out the field in which the error / validation failed due to security implications / hints it provides

only applies to a subset of systems which do not have any public registration page.
